lib/spack/llnl/util/lang.py

```diff
@@ -257,10 +257,28 @@ def setter(name, value):
 
 
 @key_ordering
-class HashableMap(dict):
+class HashableMap(collections.MutableMapping):
```

Comment:
@tgamblin @scheibelp @becker33 The change here should be a more idiomatic way to define a custom mutable mapping. Not deriving from `dict` prevents inconsistencies among different methods, like `__getitem__` and `get`, if one forgets to override both. On the other side delegating to `dict` instead of deriving from it comes with a performance cost that is non negligible for certain operations:
```console
# current develop
$ time spack spec -Il trilinos
Input spec
...
real	0m3.177s
user	0m3.088s
sys	0m0.072s

# This branch
$ time spack spec -Il trilinos
Input spec
...
real	0m3.812s
user	0m3.704s
sys	0m0.088s
```
Let me know if you want me to revert this.